beets/autotag/hooks.py

```diff
@@ -214,16 +230,32 @@ def __init__(
 
 
 # Structures that compose all the information for a candidate match.
+@dataclass
+class Match:
+    distance: Distance
+    info: Info
 
+    @cached_classproperty
+    def type(cls) -> str:
+        return cls.__name__.removesuffix("Match")  # type: ignore[attr-defined]
 
-class AlbumMatch(NamedTuple):
-    distance: Distance
+
+@dataclass
+class AlbumMatch(Match):
     info: AlbumInfo
     mapping: dict[Item, TrackInfo]
     extra_items: list[Item]
     extra_tracks: list[TrackInfo]
 
+    @property
+    def item_info_pairs(self) -> list[tuple[Item, TrackInfo]]:
+        return list(self.mapping.items())
 
-class TrackMatch(NamedTuple):
-    distance: Distance
+    @property
+    def items(self) -> list[Item]:
```

Comment:
This should be `self.mapping.keys()` instead no need to loop over the `item_info_pairs`.

Reply:
It is still the main data container right now for the pairs, I'm open to changing this in general at some point but this needs a bit more thought in my opinion. Maybe properties for extra_* and pairs might be an elegant way forward with one candidates data container and index mappings for the properties.  How I think about this is kinda like a venn diagram.
```
// items ∩ tracks = pairs
// items' ∩ tracks = extra_items
// items ∩ tracks' = extra_tracks
```
So all these properties are derived from the `items` and `tracks` with some sort of index mapping.